rslib/src/search/sqlwriter.rs

```diff
@@ -419,57 +421,119 @@ impl SqlWriter<'_> {
+        self.args.push(format!("(?i)^{}$", to_re(val)));
+        write!(self.sql, "regexp_fields(?{}, n.flds)", self.args.len()).unwrap();
+    }
 
+    fn write_single_field_regexp(&mut self, field_name: &str, val: &str) -> Result<()> {
+        let field_map = self.field_map(field_name)?;
         if field_map.is_empty() {
             write!(self.sql, "false").unwrap();
             return Ok(());
         }
 
-        let cmp;
-        let cmp_trailer;
-        if is_re {
-            cmp = "regexp";
-            cmp_trailer = "";
-            self.args.push(format!("(?i){}", val));
-        } else {
-            cmp = "like";
-            cmp_trailer = "escape '\\'";
-            self.args.push(to_sql(val).into())
+        self.args.push(format!("(?i){}", val));
+        let arg_idx = self.args.len();
+        write!(
+            self.sql,
+            "({})",
+            field_map
+                .iter()
+                .map(|(mid, fields)| {
+                    format!(
+                        "(n.mid = {} and regexp_fields(?{}, n.flds, {}))",
+                        mid,
+                        arg_idx,
+                        fields.iter().join(", "),
+                    )
+                })
+                .join(" or ")
+        )
+        .unwrap();
+
+        Ok(())
+    }
+
+    fn write_single_field(&mut self, field_name: &str, val: &str) -> Result<()> {
+        let field_map = self.field_map(field_name)?;
+        if field_map.is_empty() {
+            write!(self.sql, "false").unwrap();
+            return Ok(());
         }
 
+        self.args.push(to_sql(val).into());
         let arg_idx = self.args.len();
-        let searches: Vec<_> = field_map
-            .iter()
-            .map(|(ntid, ord)| {
-                format!(
-                    "(n.mid = {mid} and field_at_index(n.flds, {ord}) {cmp} ?{n} {cmp_trailer})",
-                    mid = ntid,
-                    ord = ord.unwrap_or_default(),
-                    cmp = cmp,
-                    cmp_trailer = cmp_trailer,
-                    n = arg_idx
-                )
-            })
-            .collect();
-        write!(self.sql, "({})", searches.join(" or ")).unwrap();
+        write!(
+            self.sql,
+            "({})",
+            field_map
+                .iter()
+                .map(|(mid, fields)| {
+                    format!(
+                        "(n.mid = {} and ({}))",
+                        mid,
+                        fields
+                            .iter()
+                            .map(|ord| {
+                                format!(
+                                    "field_at_index(n.flds, {}) like ?{} escape '\\'",
+                                    ord, arg_idx
+                                )
+                            })
+                            .join(" or "),
+                    )
+                })
+                .join(" or ")
+        )
+        .unwrap();
```

Comment:
This is getting a bit unwieldy with the nesting; it may make sense to declare some of the closures above the write!() statement to break it up a bit. Now that we're on Rust 1.58, you can also take advantage of inline replacements: 
https://blog.rust-lang.org/2022/01/13/Rust-1.58.0.html#captured-identifiers-in-format-strings